File: src/main/java/mediathek/gui/MVTray.java
package mediathek.gui;

import mediathek.config.Daten;
import mediathek.config.Icons;
import mediathek.daten.DownloadStartInfo;
import mediathek.gui.messages.TimerEvent;
import mediathek.gui.messages.TrayIconEvent;
import mediathek.mainwindow.MediathekGui;
import mediathek.tool.ApplicationConfiguration;
import mediathek.tool.notification.thrift.MessageType;
import mediathek.tool.notification.thrift.NotificationMessage;
import net.engio.mbassy.listener.Handler;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import javax.swing.*;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;

public final class MVTray {

    private final Daten daten;
    private int trayState; // 0, 1=Download, 2=Download mit Fehler
    private SystemTray tray;
    private TrayIcon trayIcon;
    private int count;

    public MVTray() {
        daten = Daten.getInstance();
        daten.getMessageBus().subscribe(this);
    }

    @Handler
    private void handleTimerEvent(TimerEvent msg) {
        SwingUtilities.invokeLater(() -> {
            ++count;
            if (count > 3) {
                // nur alle 3s ändern
                trayIcon.setToolTip(getInfoTextDownloads());
                count = 0;
            }

            // Anzahl, Anz-Abo, Anz-Down, nicht gestarted, laufen, fertig OK, fertig fehler
            DownloadStartInfo info = daten.getListeDownloads().getStarts();
            if (info.error > 0) {
                // es gibt welche mit Fehler
                if (trayState != 2) {
                    trayState = 2;
                    trayIcon.setImage(Icons.ICON_TRAY_ERROR);
                }
            } else if (info.running > 0) {
                // es laufen welche
                if (trayState != 1) {
                    trayState = 1;
                    trayIcon.setImage(Icons.ICON_TRAY_DOWNLOAD);
                }
            } else if (trayState != 0) {
                trayState = 0;
                trayIcon.setImage(Icons.ICON_TRAY);
            }
        });
    }

    public void beenden() {
        if (tray != null && trayIcon != null) {
            tray.remove(trayIcon);
        }
    }

    public MVTray systemTray() {
        if (!SystemTray.isSupported()) {
            logger.info("Tray wird nicht unterstützt");
            return null;
        } else {
            tray = SystemTray.getSystemTray();
            trayIcon = new TrayIcon(Icons.ICON_TRAY);
            trayIcon.setImageAutoSize(true);
            trayIcon.setToolTip(getInfoTextDownloads());
            addListener();

            final PopupMenu popup = new PopupMenu();
            trayIcon.setPopupMenu(popup);

            MenuItem itemInfo = new MenuItem("Infos anzeigen");
            itemInfo.addActionListener(e -> addNotification(getTextInfos()));
            popup.add(itemInfo);

            MenuItem itemRemoveTray = new MenuItem("Trayicon ausblenden");
            itemRemoveTray.addActionListener(e -> {
                MediathekGui.ui().setVisible(true);
                ApplicationConfiguration.getConfiguration().setProperty(ApplicationConfiguration.APPLICATION_UI_USE_TRAY,false);
                MediathekGui.ui().initializeSystemTray();
                daten.getMessageBus().publishAsync(new TrayIconEvent());
            });
            popup.add(itemRemoveTray);

            popup.addSeparator();
            MenuItem itemBeenden = new MenuItem("Programm beenden");
            itemBeenden.addActionListener(e -> MediathekGui.ui().beenden(false, false));
            popup.add(itemBeenden);

            trayIcon.setPopupMenu(popup);
            try {
                tray.add(trayIcon);
                return this;
            } catch (AWTException e) {
                logger.error("Tray konnte nicht geladen werden", e);
            }

        }
        return null;
    }

    private static final Logger logger = LogManager.getLogger(MVTray.class);

    private void addListener() {
        trayIcon.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
                if (e.getButton() == MouseEvent.BUTTON1) {
                    if (e.getClickCount() == 1) {
                        final MediathekGui ui = MediathekGui.ui();
                        ui.setVisible(!ui.isVisible());
                        if (ui.isVisible()) {
                            ui.toFront();
                            ui.requestFocus();
                        }
                    }
                }
            }
        });
    }

    private String getTextInfos() {
        String strText = "";

        strText += "Filmliste erstellt: " + daten.getListeFilme().metaData().getGenerationDateTimeAsString() + " Uhr  ";
        strText += "\n";
        strText += "Anz. Filme: " + daten.getListeFilme().size();
        strText += "\n";
        strText += getInfoTextDownloads();

        return strText;
    }

    private String getInfoTextDownloads() {
        final DownloadStartInfo info = daten.getListeDownloads().getStarts();
        String text = "Downloads: " + info.total_starts;

        if (info.hasValues()) {
            text += "   [ ";
            text += (info.running == 1) ? "1 läuft" : info.running + " laufen";

            if (info.running > 0)
                text += " (" + daten.getDownloadInfos().getBandwidthStr() + ')';

            text += (info.initialized == 1) ? ", 1 wartet" : ", " + info.initialized + " warten";

            if (info.finished > 0)
                text += (info.finished == 1) ? ", 1 fertig" : ", " + info.finished + " fertig";

            if (info.error > 0)
                text += (info.error == 1) ? ", 1 fehlerhaft" : ", " + info.error + " fehlerhaft";

            text += " ]";
        }

        return text;
    }

    private void addNotification(String meldung) {
        final NotificationMessage msg = new NotificationMessage();
        msg.setTitle("Programminfos");
        msg.setMessage(meldung);
        msg.setType(MessageType.INFO);
        daten.notificationCenter().displayNotification(msg);
    }

}

File: src/main/kotlin/mediathek/gui/MVTray.kt
package mediathek.gui

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.swing.Swing
import mediathek.config.Daten
import mediathek.config.Konstanten
import mediathek.config.application.ApplicationConfiguration
import mediathek.gui.messages.TimerEvent
import mediathek.gui.messages.TrayIconEvent
import mediathek.mainwindow.TrayHost
import mediathek.tool.GetIcon
import mediathek.tool.MessageBus
import mediathek.tool.notification.MessageType
import mediathek.tool.notification.NotificationMessage
import mediathek.tool.notification.NotificationService
import net.engio.mbassy.listener.Handler
import org.apache.logging.log4j.LogManager
import java.awt.AWTException
import java.awt.MenuItem
import java.awt.PopupMenu
import java.awt.SystemTray
import java.awt.TrayIcon
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent

class MVTray(
    private val host: TrayHost,
) {
    private enum class TrayState {
        IDLE,
        DOWNLOADING,
        ERROR,
    }

    private var trayState = TrayState.IDLE
    private var tray: SystemTray? = null
    private var trayIcon: TrayIcon? = null
    private var count = 0
    private var subscribed = false
    private var uiScope = createUiScope()

    @Handler
    @Suppress("UNUSED_PARAMETER")
    private fun handleTimerEvent(msg: TimerEvent) {
        uiScope.launch {
            val currentTrayIcon = trayIcon ?: return@launch

            ++count
            if (count > 3) {
                // nur alle 3s ändern
                currentTrayIcon.toolTip = infoTextDownloads
                count = 0
            }

            // Anzahl, Anz-Abo, Anz-Down, nicht gestarted, laufen, fertig OK, fertig fehler
            val info = Daten.getInstance().listeDownloads.starts
            if (info.error > 0) {
                // es gibt welche mit Fehler
                if (trayState != TrayState.ERROR) {
                    trayState = TrayState.ERROR
                    currentTrayIcon.image = GetIcon.getProgramIcon("tray-fehler.png", 256, 256).image
                }
            } else if (info.running > 0) {
                // es laufen welche
                if (trayState != TrayState.DOWNLOADING) {
                    trayState = TrayState.DOWNLOADING
                    currentTrayIcon.image = GetIcon.getProgramIcon("tray-download.png", 256, 256).image
                }
            } else if (trayState != TrayState.IDLE) {
                trayState = TrayState.IDLE
                currentTrayIcon.image = Konstanten.ICON_TRAY
            }
        }
    }

    fun beenden() {
        val currentTray = tray
        val currentTrayIcon = trayIcon
        if (currentTray != null && currentTrayIcon != null) {
            currentTray.remove(currentTrayIcon)
        }
        if (subscribed) {
            MessageBus.messageBus.unsubscribe(this)
            subscribed = false
        }
        uiScope.cancel()
        tray = null
        trayIcon = null
    }

    fun systemTray(): MVTray? {
        if (!SystemTray.isSupported()) {
            logger.info("Tray wird nicht unterstützt")
            return null
        }

        if (!uiScope.isActive) {
            uiScope = createUiScope()
        }

        val newTray = SystemTray.getSystemTray()
        val newTrayIcon = TrayIcon(Konstanten.ICON_TRAY)
        tray = newTray
        trayIcon = newTrayIcon
        newTrayIcon.isImageAutoSize = true
        newTrayIcon.toolTip = infoTextDownloads
        addListener(newTrayIcon)

        val popup = PopupMenu()
        newTrayIcon.popupMenu = popup

        val itemInfo = MenuItem("Infos anzeigen")
        itemInfo.addActionListener { addNotification(textInfos) }
        popup.add(itemInfo)

        val itemRemoveTray = MenuItem("Trayicon ausblenden")
        itemRemoveTray.addActionListener {
            host.showMainWindow()
            ApplicationConfiguration.getInstance().useTray = false
            host.refreshSystemTray()
            MessageBus.messageBus.publishAsync(TrayIconEvent())
        }
        popup.add(itemRemoveTray)

        popup.addSeparator()
        val itemBeenden = MenuItem("Programm beenden")
        itemBeenden.addActionListener { host.quitApplication() }
        popup.add(itemBeenden)

        newTrayIcon.popupMenu = popup
        try {
            newTray.add(newTrayIcon)
            MessageBus.messageBus.subscribe(this)
            subscribed = true
            return this
        } catch (e: AWTException) {
            logger.error("Tray konnte nicht geladen werden", e)
            tray = null
            trayIcon = null
        }

        return null
    }

    private fun addListener(currentTrayIcon: TrayIcon) {
        currentTrayIcon.addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(e: MouseEvent) {
                if (e.button == MouseEvent.BUTTON1 && e.clickCount == 1) {
                    host.toggleMainWindowVisibility()
                }
            }
        })
    }

    private val textInfos: String
        get() {
            val filmList = Daten.getInstance().listeFilme
            return buildString {
                append("Filmliste erstellt: ")
                append(filmList.metaData.generationDateTimeAsString)
                append(" Uhr  \n")
                append("Anz. Filme: ")
                append(filmList.size)
                append('\n')
                append(infoTextDownloads)
            }
        }

    private val infoTextDownloads: String
        get() {
            val daten = Daten.getInstance()
            val info = daten.listeDownloads.starts
            return buildString {
                append("Downloads: ")
                append(info.total_starts)

                if (info.hasValues()) {
                    append("   [ ")
                    append(if (info.running == 1) "1 läuft" else "${info.running} laufen")

                    if (info.running > 0) {
                        append(" (")
                        append(daten.downloadInfos.bandwidthStr)
                        append(')')
                    }

                    append(if (info.initialized == 1) ", 1 wartet" else ", ${info.initialized} warten")

                    if (info.finished > 0) {
                        append(if (info.finished == 1) ", 1 fertig" else ", ${info.finished} fertig")
                    }

                    if (info.error > 0) {
                        append(if (info.error == 1) ", 1 fehlerhaft" else ", ${info.error} fehlerhaft")
                    }

                    append(" ]")
                }
            }
        }

    private fun addNotification(meldung: String) {
        val msg = NotificationMessage()
        msg.title = "Programminfos"
        msg.message = meldung
        msg.type = MessageType.INFO
        NotificationService.displayNotification(msg)
    }

    companion object {
        private val logger = LogManager.getLogger()

        private fun createUiScope() = CoroutineScope(SupervisorJob() + Dispatchers.Swing)
    }
}
